### PR TITLE
Namespace role filtering

### PR DESCRIPTION
**Pull request by TheBurchLog**
This is the redesign of the user roles that also will filter on the data objects based on their namespace affiliation.

TODOS:
- [x] Update Requests View Page
- [x] Filter Events
- [x] Figure out why first selected option on Roles blanks out the permissions
- [x] Test -- Admin Access to one namespace and Read Access to another, what happens on the System Admin page?
- [x] Get working with Parent HTTP requests
- [x] Get working with Child HTTP requests
- [x] Test all the pages for filtering
- [x] Test out the filtering in weird use cases
- [x] Verify that the cookies work properly with the new models
- [x] Unit Tests

Round Two TODOS:
- [x] Get the Add Role buttons working
- [x] Verify Requests page filters on the Database layer for namespaces
- [x] Replace CREATE/MAINTAINER with Operator
- [x] Replace LOCAL with GARDEN in UI
- [x] Move Filter out of HTTP code
- [x] Fix Add Roles to Users

Round Three TODOS
- [x] Replace `is_local` with `garden` in models
- [x] Add user input text field for Roles Page
- [x] Redesign how tables work for Roles and Users
- [x] Create custom DB mapping for database side filters
- [x] Convert Namespace functions to return Gardens as well
- [x] Apply Garden record to all Systems at Create
- [x] Apply Garden record to all Requests at Create (according to their target System)
- [x] Update UI permissions function to account for Gardens
- [x] Update all Request APIs to expect Garden records
- [x] Update Request logic to check for Garden/System mapping before executing
- [ ] Add new Migration scripts

Round Four TODOs
- [x] ** In Proxy PR, add role in memory role model based off string parser of role name **
- [x] Update Admin Garden UI to filter on Garden for Systems instead of using internal System Routing list
- [x] Make none route-able garden objects for the presentation layer for Admin Garden UI (based on `$rootScope.systems`)

Round Five TODOs
- [x] Display what Garden's are nested on the routing gardens page

New Filtering Decisions
- If Namespace and Garden is set in the same permission, then the user get the intersect of those permissions.
- If the Namespace or Garden is set, then the user gets the entirety of that scope.
- Permissions are the least restrictive of the combination. So if one permission gives access to Garden and another gives access to Namespace, then the user gets both, unlike the first decision point.
- Admin access on the Local garden gives the user the ability to manage anything within that Garden space that is not tied to Garden/Namespace (Users/Roles/Rescans)

brewtils: namespace-role-filtering

Depends on https://github.com/beer-garden/brewtils/pull/274
Depends on https://github.com/beer-garden/angular-schema-form-builder/pull/17

### CODE REVIEW

**Comment by SpecificallyGravity** [comment]
@TheBurchLog In reference to the UI blanking issue:

It could be how `ng-repeat` and `ng-model` interact as seen [on stackoverflow](https://stackoverflow.com/questions/14049480/what-are-the-nuances-of-scope-prototypal-prototypical-inheritance-in-angularjs). That being said, as far as I can tell you're still doing things correctly.  I'm able to fix the blanking issue if I change `ng-model` to do something that throws an error and I couldn't begin to understand why.

**Comment by TheBurchLog** [comment]
@SpecificallyGravity So the solution I am going with on the `ng-repeat` and `ng-model` scenario is to reload the models on the first selection. Not pretty, but it gets around the issue for now.

**Comment by hazmat345** [comment]
@TheBurchLog I'm starting to look at this, but this is a 3k diff. Can you please sketch out an overview of what this does / how it does it?

**Comment by hazmat345** [comment]
At a high level - it looks like all the auth stuff is tightly integrated with the `http` module. How will all of that be applied to alternate entry points?

**Comment by TheBurchLog** [comment]
I am going to work on simplifying the accesses a step further.

READ - Has access to view data stored in the scope assigned
OPERATOR - Has access to create/edit/delete data in the scope assigned
ADMIN - Has access to manage data in the scope assigned

File: src/app/beer_garden/router.py [hunk 1]
```diff
@@ -610,29 +652,25 @@ def _forward_http(operation: Operation, target_garden: Garden):
...
+    easy_client = EasyClient(
+        bg_host=conn_info.get("host"),
+        bg_port=conn_info.get("port"),
+        ssl_enabled=conn_info.get("ssl"),
+        bg_url_prefix=conn_info.get("url_prefix", "/"),
+        username=conn_info.get("username", None),
+        password=decoded_password,
+        ca_cert=conn_info.get("ca_cert"),
+        ca_verify=conn_info.get("ca_verify"),
+        client_cert=conn_info.get("client_cert"),
```
Comment (hazmat345) [inline src/app/beer_garden/router.py:669]: :+100:

File: src/app/beer_garden/router.py [hunk 2]
```diff
@@ -647,7 +685,7 @@ def _forward_http(operation: Operation, target_garden: Garden):
         elif target_garden.status != "RUNNING":
             beer_garden.garden.update_garden_status(target_garden.name, "RUNNING")
 
-        return response
+        return response.json()
```
Comment (hazmat345) [inline src/app/beer_garden/router.py:688]: Is this what we want to be returning?

**Review by hazmat345 — COMMENTED** [review]
Added comments. I'll need to play with this some. But overall I think it looks good 👍🏼

File: src/app/beer_garden/db/mongo/util.py [hunk 3]
```diff
@@ -360,15 +372,37 @@ def _update_request_has_parent_model():
...
     """Create a role if it doesn't already exist"""
+
     from .models import Role
 
     try:
         Role.objects.get(name=role.name)
     except DoesNotExist:
         logger.warning("Role %s missing, about to create" % role.name)
         role.save()
+        logger.info("Added %s" % role.name)
```
[the remark this answers is not in the document]
Comment (TheBurchLog) [reply src/app/beer_garden/db/mongo/util.py:405]: Nope, this was from all that debugging...

File: src/app/beer_garden/api/stomp/authorization.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
+# -*- coding: utf-8 -*-
+import base64
+
+import jwt
+
+from beer_garden.users import coalesce_permissions
+from brewtils.models import (
+    Principal as BrewtilsPrincipal,
+    Role as BrewtilsRole,
+    Permission as BrewtilsPermission,
```
[the remark this answers is not in the document]
Comment (TheBurchLog) [reply src/app/beer_garden/api/stomp/authorization.py:10]: It is 100% that file. This was future proofing for if we wanted to use authorization for STOMP entry points. We can roll this out until it is needed.

**Comment by TheBurchLog** [comment]
Testing Strategy for this:

Deployed
| ID | Garden | Namespace | System |
| --- | --- | --- | --- |
| 1 | Default | Test-1 | Echo |
| 2 | Default | Test-2 | Sleep |
| 3 | Child | Test-2 | Complex|

Permission Access

| ID | Garden | Namespace | Access | 
| --- | --- | --- | --- | 
| 1 | Default | - | Admin | 
| 2 | Default | Test-1 | Admin | 
| 3 | - | Test-2 | READ | 
| 4 | - | Test-1 | READ | 
| 5 | Child | - | ADMIN | 
| 6 | - | Test-2 | ADMIN | 
 

Permission Mixing & Results
| Permission IDs | Read Systems | Admin Systems | Admin Users |
| --- | --- | --- | --- |
| 1 | 1,2,3 | 1,2,3 | True |
| 2 | 1 | 1 | False |
| 3 | 2,3 | - | False |
| 4 | 1 | - | False |
| 5 | 3 | 3 | False |
| 6 | 2,3 | 2,3 | False |
| 4,5 | 1,2,3 | 3 | False |
| 3,5 | 2,3 | 3 | False |
| 2,6 | 1,2,3 | 1,2,3 | False |

**Comment by hazmat345** [comment]
@TheBurchLog trying to play with this and I'm having some trouble. I'm using the new branches for this, brewtils, and the builder and I'm using a clean database.

When I go to the main page I don't see any nav menu items:

image

- Is this because I don't have read permissions? I have `enabled` and `guest_login_enabled` both set to true in the config.
- If yes, should be just grey (disable) things instead? It's pretty disorienting to not have anything in the nav bar 😄

**Comment by TheBurchLog** [comment]
@hazmat345 

I don't remember testing `guest_login_enabled`. I think that is preventing the Login prompt. I honestly don't know what the menu is not appearing. I will have to test that on my side and see if I can replicate it.

**Pull request closed without merging**